### PR TITLE
feat: add IncomingRequest::is() method

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
Writing code like `strtolower($this->request->getMethod()) !== 'post'` is a pain.

- add `IncomingRequest::is()`

```php
$request->is('get')
$request->is('post')
$request->is('put')
$request->is('delete')
$request->is('head')
$request->is('patch')
$request->is('options')
$request->is('json')
$request->is('ajax')
```

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Review by datamweb — COMMENTED** [review]
To be beautiful code.😄

File: system/HTTP/IncomingRequest.php [hunk 1]
```diff
@@ -375,6 +375,29 @@ public function negotiate(string $type, array $supported, bool $strictMatch = fa
...
+    {
+        $value = strtolower($value);
+
+        $httpMethods = ['get', 'post', 'put', 'delete', 'head', 'patch', 'options'];
+
+        if (in_array($value, $httpMethods, true)) {
+            return strtolower($this->getMethod()) === $value;
+        }
+
+        if ($value === 'json') {
```
Comment (michalsn) [inline system/HTTP/IncomingRequest.php:394]: Not sure if `json` belong here. But if we will have it, why not also have `ajax`? Basically the same story.

**Comment by iRedds** [comment]
1. 
```php
public function isMethod(string $method) : bool {}
public function isGet() : bool {}
public function isPost() : bool {}
// e.t.c.
```
2. the default method name is in uppercase.
The fact that in the framework the method name is in lowercase is the wrong way.

File: system/HTTP/IncomingRequest.php (hunk 1 above)
Comment (kenjis) [reply system/HTTP/IncomingRequest.php:394]: I'm fine to add `ajax`.

**Comment by kenjis** [comment]
> The fact that in the framework the method name is in lowercase is the wrong way.

What do you mean? Why is lowercase wrong way?

File: system/HTTP/IncomingRequest.php (hunk 1 above)
Comment (kenjis) [reply system/HTTP/IncomingRequest.php:394]: Added.

Comment (iRedds) [reply system/HTTP/IncomingRequest.php:394]: We already have the isAJAX() method.

**Comment by MGatner** [comment]
I think @iRedds was pointing out that we only need this method because we are inconsistent with our case in the `getMethod()` return. The `$upper` parameter is already deprecated which means we are currently forcing lowercase method names. However if we want to go PSR-7 we cannot modify the case:
```
     * While HTTP method names are typically all uppercase characters, HTTP
     * method names are case-sensitive and thus implementations SHOULD NOT
     * modify the given string.
```

I'm not sure the best approach here. I am not opposed to this helper method, but I also believe it should be unnecessary.

**Comment by michalsn** [comment]
The lowercase/uppercase "problem" was indeed created by PSR-7 compatibility. I don't see a way to resolve it without introducing quite a big BC break because, by default, everything is in lowercase now (and was the same way in CI3).

File: system/HTTP/IncomingRequest.php (hunk 1 above)
Comment (michalsn) [reply system/HTTP/IncomingRequest.php:394]: Yes, we have, and as you may notice, we use it. IMO proposed `is()` method would serve as a convenient shortcut, so I don't see a problem with another way of accessing the same method.

**Comment by kenjis** [comment]
> Because "the default method name is in uppercase."

> The method token is case-sensitive because it might be used as a gateway to object-based systems with case-sensitive method names. By convention, standardized methods are defined in all-uppercase US-ASCII letters.
https://www.rfc-editor.org/rfc/rfc9110#name-overview

It seems method names should be case-sensitive.

The `getMethod()` returns lowercase now.
So `strtolower($this->getMethod()) === $value;` is redundant.
But the default `$upper = false` in `getMethod()` will be removed.

**Comment by kenjis** [comment]
@iRedds Do you say that the standard HTTP method names are in uppercase?
I changed the implementation a bit.

```console
$ curl -X get https://www.w3.org/ -i
HTTP/2 400 
server: cloudflare
date: Wed, 21 Dec 2022 08:40:47 GMT
content-type: text/html
content-length: 155
cf-ray: -

<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
```

**Comment by MGatner** [comment]
I think case-sensitivity is in our favor. We don't need to adjust the case on the method return because `Get` is not `GET`. Forcing these to lower does appear to be a violation of the spec; maybe we add a new method for now?

**Comment by kenjis** [comment]
> The lowercase/uppercase "problem" was indeed created by PSR-7 compatibility. I don't see a way to resolve it without introducing quite a big BC break because, by default, everything is in lowercase now (and was the same way in CI3).

Yes, the current `getMethod()` returns values that are converted to lowercase, and it seems a violation of the HTTP specification. 

The behavior should be fixed but it is a big BC break. So it is better that we wait for v5.0.

@MGatner This `is()` method is the new method to check HTTP methods in case insensitive manner.

**Comment by michalsn** [comment]
> The behavior should be fixed but it is a big BC break. So it is better that we wait for v5.0.

I agree with this.

`is()` may serve as a handful helper method, and since it handles `json` and `ajax` it's not 100% redundant with `getMethod()`.

**Review by michalsn — APPROVED** [review]

**Comment by kenjis** [comment]
Rebased and resolved, and all green!

**Comment by kenjis** [comment]
Added docs.

File: user_guide_src/source/changelogs/v4.3.0.rst [hunk 2]
```diff
@@ -323,6 +323,8 @@ Others
 - **Config:** Added ``Config\Session`` class to handle session configuration.
 - **Debug:** Kint has been updated to 5.0.1.
 - **Request:** Added new ``$request->getRawInputVar()`` method to return a specified variable from raw stream. See :ref:`Retrieving Raw data <incomingrequest-retrieving-raw-data>`.
+- **Request:** Added new ``$request->is()`` method to to query the request type.
```
Comment (datamweb) [inline user_guide_src/source/changelogs/v4.3.0.rst:326]:
```suggestion
- **Request:** Added new ``$request->is()`` method to query the request type.
```

**Review by MGatner — APPROVED** [review]
I do t feel great about it but I agree with the logic that landed here. Definitely we should fix all this in v5. 

Also, we need to move away from our AJAX methods as they are less and less reliable; maybe in the future we can provide some different terms for the same concept.

**Comment by michalsn** [comment]
> next major version?

@MGatner I'm not sure if this label was meant to be added to this PR, but if so, I would have second thoughts about this PR. v5 means that we already have the `getMethod()` fixed, so `is()` method wouldn't make as much sense to me. 

Then, I would rather add the `isJSON()` method.

**Comment by MGatner** [comment]
@michalsn I've been using that label for PRs or threads I want to come back and review for v5 content. It doesn't necessarily mean "this PR belongs in v5" (though sometimes) but more like "read this thread for a relevant v5 idea".

**Comment by kenjis** [comment]
Rebased and pushed two commits.

**Review by MGatner — APPROVED** [review]

**Comment by kenjis** [comment]
> Then, I would rather add the `isJSON()` method.

I found `isJSON()` method. It was deprecated in 4.0.5.
See https://github.com/codeigniter4/CodeIgniter4/pull/7019/files#diff-2cab244a084ce08ac666dff9a88aee13003b4170f6c02a690263e3260dac26d8R35

And it was removed in 4.1.2.
See https://github.com/codeigniter4/CodeIgniter4/commit/55b43083

**Comment by michalsn** [comment]
Well... it's strange to me why we deprecated `isJSON()`... and now we're, in fact, introducing it again. I tried to follow the related discussions, but I still have no point. Anyway, I don't mind it at all. This is a common functionality that will make the life of developers easier. At least, in my opinion.

**Comment by MGatner** [comment]
I think `isJSON()` was removed when we were trying to establish PSR-7 compatibility, but before we split framework-specific methods out into traits. At least that's my best recollection. It's convenient, I have no problem with it, but it is one more framework-coupling method that doesn't save on lines of code or much readability.

**Pull request merged**